crates/rattler_installs_packages/src/wheel_builder/mod.rs

```diff
@@ -227,19 +254,119 @@ impl<'db, 'i> WheelBuilder<'db, 'i> {
     /// Build a wheel from an sdist by using the build_backend in a virtual env.
     /// This function uses the `build_wheel` entry point of the build backend.
     #[tracing::instrument(skip_all, fields(name = %sdist.name().distribution.as_source_str(), version = %sdist.name().version))]
-    pub async fn build_wheel(&self, sdist: &SDist) -> Result<PathBuf, WheelBuildError> {
+    pub async fn build_wheel(&self, sdist: &SDist) -> Result<Wheel, WheelBuildError> {
+        // Check if we have already built this wheel locally and use that instead
+        let key = WheelKey::try_from(sdist)?;
+        if let Some(wheel) = self.locally_built_wheels.wheel_for_key(&key)? {
+            return Ok(wheel);
+        }
+
+        // Setup a new virtualenv for building the wheel or use an existing
         let build_environment = self.setup_build_venv(sdist).await?;
 
+        // Run the wheel stage
         let output = build_environment.run_command("Wheel")?;
 
+        // Check for success
         if !output.status.success() {
             let stdout = String::from_utf8_lossy(&output.stderr);
             return Err(WheelBuildError::Error(stdout.to_string()));
         }
 
-        let result = std::fs::read_to_string(build_environment.work_dir().join("wheel_result"))?;
-        let wheel_file = PathBuf::from(result.trim());
+        // This is where the wheel file is located
+        let wheel_file: PathBuf =
+            std::fs::read_to_string(build_environment.work_dir().join("wheel_result"))?
+                .trim()
+                .into();
+
+        // Get the name of the package
+        let package_name: NormalizedPackageName = sdist.name().distribution.clone().into();
+
+        // Save the wheel into the cache
+        let key = WheelKey::try_from(sdist)?;
+
+        // Reconstruction of the wheel filename
+        let file_component = wheel_file
+            .file_name()
+            .and_then(|f| f.to_str())
+            .ok_or_else(|| {
+                WheelBuildError::Error(format!(
+                    "Could not get extract file component from {}",
+                    wheel_file.display()
+                ))
+            })?;
+        let wheel_file_name = WheelFilename::from_filename(file_component, &package_name)?;
+
+        // Associate the wheel with the key which is the hashed sdist
+        self.locally_built_wheels.associate_wheel(
+            &key,
+            wheel_file_name,
+            &mut std::fs::File::open(&wheel_file)?,
+        )?;
+
+        // Reconstruct wheel from the path
+        let wheel = Wheel::from_path(&wheel_file, &package_name).map_err(|e| {
+            WheelBuildError::Error(format!(
+                "Could not build wheel for metadata extraction: {}",
```

Comment:
is this the right error message? Because at this point we're building the wheel for installation, right?